### PR TITLE
Abi 2.1

### PR DESCRIPTION
**Pull request by AlexeyVavilin**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ton_abi"
-version = "1.4.0"
+version = "1.4.1"
```
Comment (melsomino) [inline Cargo.toml:3]: May be set version to 2.1.0 as an ABI version (or at least 1.5.0)

**Review by melsomino — CHANGES_REQUESTED** [review]
See comments

File: src/token/deserialize.rs [hunk 2]
```diff
@@ -51,29 +52,55 @@ impl TokenValue {
...
+    }
+
+    fn read_int(size: usize, cursor: SliceData) -> Result<(Self, SliceData)> {
+        let (number, cursor) = Self::read_int_from_chain(size, cursor)?;
         Ok((TokenValue::Int(Int { number, size }), cursor))
     }
 
+    fn read_varuint(size: usize, cursor: SliceData) -> Result<(Self, SliceData)> {
+        let (len, cursor) = Self::read_uint_from_chain(ParamType::varint_size_len(size), cursor)?;
+        let len = len.to_usize().unwrap();
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/token/deserialize.rs:92]: `len` can not be bigger then `u8`. So `usize` definitely can contain the value

File: src/contract.rs [hunk 3]
```diff
@@ -11,18 +11,61 @@
...
 use ton_block::Serializable;
 use ton_types::{BuilderData, error, fail, HashmapE, Result, SliceData};
 
-pub const SUPPORTED_VERSIONS: [u8; 2] = [1, 2];
+
+const MIN_SUPPORTED_VERSION: u32 = 1000;
+const MAX_SUPPORTED_VERSION: u32 = 2001;
+
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct AbiVersion {
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/contract.rs:32]: Done

File: src/contract.rs [hunk 4]
```diff
@@ -11,18 +11,61 @@
...
 use std::io;
 use std::collections::HashMap;
 use serde::de::{Error as SerdeError};
 use serde_json;
 use ton_block::Serializable;
 use ton_types::{BuilderData, error, fail, HashmapE, Result, SliceData};
 
-pub const SUPPORTED_VERSIONS: [u8; 2] = [1, 2];
+
+const MIN_SUPPORTED_VERSION: u32 = 1000;
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/contract.rs]: Done

File: src/param_type/param_type.rs [hunk 5]
```diff
@@ -149,4 +155,79 @@ impl ParamType {
...
+    }
+
+    pub(crate) fn is_large_optional(&self) -> bool {
+        self.max_bit_size() >= BuilderData::bits_capacity() || 
+        self.max_refs_count() >= BuilderData::references_capacity()
+    }
+
+    pub(crate) fn max_refs_count(&self) -> usize {
+        match self {
+            ParamType::Uint(_) => 0,
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/param_type/param_type.rs]: Done

File: src/token/deserialize.rs [hunk 6]
```diff
@@ -51,29 +52,55 @@ impl TokenValue {
             }
             ParamType::Bytes => Self::read_bytes(None, cursor, last, abi_version),
             ParamType::FixedBytes(size) => Self::read_bytes(Some(*size), cursor, last, abi_version),
+            ParamType::String => Self::read_string(cursor, last, abi_version),
             ParamType::Gram => {
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/token/deserialize.rs]: Done

File: src/token/deserialize.rs [hunk 7]
```diff
@@ -180,6 +207,20 @@ impl TokenValue {
         }
     }
 
+    fn read_string(cursor: SliceData, last: bool, abi_version: u8) -> Result<(Self, SliceData)> {
+        let (bytes, cursor) = Self::read_bytes(None, cursor, last, abi_version)?;
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/token/deserialize.rs]: Done

File: src/token/deserialize.rs [hunk 8]
```diff
@@ -200,6 +241,26 @@ impl TokenValue {
         }
     }
 
+    fn read_optional(param_type: &ParamType, cursor: SliceData, last: bool, abi_version: u8) -> Result<(Self, SliceData)> {
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/token/deserialize.rs]: Done

File: src/token/tokenizer.rs [hunk 9]
```diff
@@ -127,7 +128,7 @@ impl Tokenizer {
     ) -> Result<TokenValue> {
         let vec = Self::read_array(param, value)?;
```
[the remark this answers is not in the document]
Comment (AlexeyVavilin) [reply src/token/tokenizer.rs]: Done

File: Cargo.toml (hunk 1 above)
Comment (AlexeyVavilin) [reply Cargo.toml]: Done

**Review by melsomino — APPROVED** [review]

**Pull request merged**